numerics/quadrature_body.hpp

```diff
@@ -24,35 +34,133 @@ Primitive<std::invoke_result_t<Function, Argument>, Argument> Gauss(
+                        double const relative_tolerance,
+                        Primitive<std::invoke_result_t<Function, Argument>,
+                                  Argument> const previous_estimate) {
+  using Result = Primitive<std::invoke_result_t<Function, Argument>, Argument>;
+  Result const estimate = ClenshawCurtis<points>(f, lower_bound, upper_bound);
+  // This is the naïve estimate mentioned in [Gen72b], p. 339.
+  double const relative_error_estimate =
+      Hilbert<Result>::Norm(previous_estimate - estimate) /
+      Hilbert<Result>::Norm(estimate);
+  // We look for an estimated relative error smaller than
+  // |relative_tolerance * points|: since the integral is computed from |points|
+  // evaluations of |f|, it will necessarily carry a relative error proportional
+  // to |points|, so it makes no sense to look for convergence beyond that.
+  if (relative_error_estimate > relative_tolerance * points) {
+    if constexpr (points > 1 << 24) {
+      LOG(FATAL) << "Too many refinements while integrating from "
+                 << lower_bound << " to " << upper_bound;
+    } else {
+      return AutomaticClenshawCurtis<points + points - 1>(
+          f, lower_bound, upper_bound, relative_tolerance, estimate);
+    }
+  }
+  return estimate;
+}
+
+template<int initial_points, typename Argument, typename Function>
+Primitive<std::invoke_result_t<Function, Argument>, Argument>
+AutomaticClenshawCurtis(
+    Function const& f,
+    Argument const& lower_bound,
+    Argument const& upper_bound,
+    double const relative_tolerance) {
+  using Result = Primitive<std::invoke_result_t<Function, Argument>, Argument>;
+  // TODO(egg): factor the evaluations of f.
+  Result const estimate =
+      ClenshawCurtis<initial_points>(f, lower_bound, upper_bound);
+  return AutomaticClenshawCurtis<initial_points + initial_points - 1>(
+      f,
+      lower_bound,
+      upper_bound,
+      relative_tolerance,
+      estimate);
+}
+
+template<int points, typename Argument, typename Function>
+Primitive<std::invoke_result_t<Function, Argument>, Argument> ClenshawCurtis(
+    Function const& f,
+    Argument const& lower_bound,
+    Argument const& upper_bound) {
+  // We follow the notation from [Gen72b] and [Gen72c].
+  using Value = std::invoke_result_t<Function, Argument>;
+
+  constexpr int N = points - 1;
+  constexpr int log2_N = FloorLog2(N);
+  static_assert(N == 1 << log2_N);
+
+  Difference<Argument> const half_width = (upper_bound - lower_bound) / 2;
+
+  constexpr Angle N⁻¹π = π * Radian / N;
+
+  // We use a discrete Fourier transform rather than a cosine transform, see
+  // [Gen72c], (3).
+  // TODO(egg): Consider a discrete cosine transform, ideally incrementally
+  // computed to improve the performance of the automatic quadrature.
+
+  // If we identify [lower_bound, upper_bound] with [-1, 1],
+  // f_cos_N⁻¹π[s] is f(cos πs/N).
+  std::vector<Value> f_cos_N⁻¹π;
+  f_cos_N⁻¹π.resize(2 * N);
```

Comment:
It would be more efficient to do `reserve` and then `push_back` in the loops: no need to zero-initialize the vector.

Reply:
Shrug; let’s remain close to the notation from the paper.